runtime/resolvers/sql.go

```diff
@@ -183,87 +207,58 @@ func (r *SQLResolver) generalExport(ctx context.Context, w io.Writer, opts *runt
 	return nil
 }
 
-func (r *SQLResolver) Close() error {
-	r.releaseFunc()
-	return nil
-}
-
-func (r *SQLResolver) generateFilename() string {
-	return "api_export_" + time.Now().Format("2006-01-02T15-04-05.000Z")
-}
-
-func resolveSQLAndDeps(ctx context.Context, sqlTemplate string, opts *runtime.APIResolverOptions) (string, []*runtimev1.ResourceName, error) {
-	dialect, err := getDialect(ctx, opts.Runtime, opts.InstanceID)
+// buildSQL resolves the SQL template and returns the resolved SQL and the resource names it references.
+func buildSQL(sqlTemplate string, dialect drivers.Dialect, args, userAttributes map[string]any, forExport bool) (string, []*runtimev1.ResourceName, error) {
+	// Extract refs from the resolved SQL
+	var refs []*runtimev1.ResourceName
+	meta, err := compilerv1.AnalyzeTemplate(sqlTemplate)
 	if err != nil {
 		return "", nil, err
 	}
-	var deps []*runtimev1.ResourceName
+	for _, ref := range meta.Refs {
+		if ref.Kind == compilerv1.ResourceKindUnspecified {
+			// We don't know if it's a source or model (or neither), so we add both. Refs are just approximate.
+			refs = append(refs,
+				&runtimev1.ResourceName{Kind: runtime.ResourceKindSource, Name: ref.Name},
+				&runtimev1.ResourceName{Kind: runtime.ResourceKindModel, Name: ref.Name},
+			)
+			continue
+		}
 
-	sql, err := compilerv1.ResolveTemplate(sqlTemplate, compilerv1.TemplateData{
-		User:       opts.UserAttributes,
-		ExtraProps: opts.Args,
-		Self:       compilerv1.TemplateResource{}, // Not defined for resolvers
-		Resolve: func(ref compilerv1.ResourceName) (string, error) {
-			return safeSQLName(ref.Name), nil
-		},
-		Lookup: func(name compilerv1.ResourceName) (compilerv1.TemplateResource, error) {
-			// TODO: Implement this, do we need to support this?
-			return compilerv1.TemplateResource{}, nil
-		},
-	})
-	if err != nil {
-		return "", nil, err
+		refs = append(refs, runtime.ResourceNameFromCompiler(ref))
 	}
 
-	if dialect == drivers.DialectDuckDB {
-		ast, err := duckdbsql.Parse(sql)
+	// For DuckDB, we can do ref inference using the SQL AST (similar to the rillv1 compiler).
+	if dialect == drivers.DialectDuckDB && !meta.UsesTemplating {
```

Comment:
It's a good point – I'm torn on this. The issue is that in reconcile, we have a rule that if templating is used, no additional inference is done. So if we do it differently here, it will be inconsistent and potentially weirder to understand for devs.

I'll add the extra DuckDB parsing back for now, and layer it on top of any `{{ ref }}` in the templating. But will also raise a discussion about this inference in general, it's a little confusing IMO.